src/deparse.c

```diff
@@ -0,0 +1,601 @@
+ * foreign server for the given relation.  If foreignrel is an upper relation,
+ * then the output targetlist can also contains expressions to be evaluated on
+ * foreign server.
+ */
+List *
+multicorn_build_tlist_to_deparse(RelOptInfo *foreignrel)
+{
+	List	   *tlist = NIL;
+	MulticornPlanState *fpinfo = (MulticornPlanState *) foreignrel->fdw_private;
+	ListCell   *lc;
+
+	/*
+	 * For an upper relation, we have already built the target list while
+	 * checking shippability, so just return that.
+	 */
+	if (IS_UPPER_REL(foreignrel))
+		return fpinfo->grouped_tlist;
+
+	/*
+	 * We require columns specified in foreignrel->reltarget->exprs and those
+	 * required for evaluating the local conditions.
+	 */
+	tlist = add_to_flat_tlist(tlist,
+							  pull_var_clause((Node *) foreignrel->reltarget->exprs,
+											  PVC_RECURSE_PLACEHOLDERS));
+	foreach(lc, fpinfo->local_conds)
+	{
+		RestrictInfo *rinfo = lfirst_node(RestrictInfo, lc);
+
+		tlist = add_to_flat_tlist(tlist,
+								  pull_var_clause((Node *) rinfo->clause,
+												  PVC_RECURSE_PLACEHOLDERS));
+	}
+
+	return tlist;
+}
+
+/*
+ * Iterate through the targets and extract relveant information needed to execute
+ * the aggregation and/or grouping on the remote data source through Python.
+ *
+ * NB: Logic here is strongly coupled to multicorn_foreign_grouping_ok(), i.e.
+ * if there is no ressortgroupref set, we automatically assume the only other
+ * option is a Aggref node type.
+ * Moreover, for the Aggref node type we assume only a single element in args
+ * (e.g. sum(column2)). In particular, this is because in
+ * multicorn_foreign_expr_walker() we don't have T_OpExpr case yet.
+ */
+void
+multicorn_extract_upper_rel_info(PlannerInfo *root, List *tlist, MulticornPlanState *fpinfo)
+{
+    ListCell *lc;
+    TargetEntry *tle;
+    Var *var;
+    Value *colname, *function;
+    Aggref *aggref;
+    StringInfo agg_key = makeStringInfo();
+
+    foreach(lc, tlist)
+    {
+        tle = lfirst_node(TargetEntry, lc);
+
+        if (tle->ressortgroupref)
+        {
+            /* GROUP BY target */
+            var = (Var *) tle->expr;
+            colname = colnameFromVar(var, root);
+
+            fpinfo->group_clauses = lappend(fpinfo->group_clauses, colname);
+            fpinfo->upper_rel_targets = lappend(fpinfo->upper_rel_targets, colname);
+        }
+        else
+        {
+            /* Aggregation target */
+            aggref = (Aggref *) tle->expr;
+            function = multicorn_deparse_function_name(aggref->aggfnoid);
+
+            var = linitial(pull_var_clause((Node *) aggref,
+                                            PVC_RECURSE_AGGREGATES |
+                                            PVC_RECURSE_PLACEHOLDERS));
+            colname = colnameFromVar(var, root);
+
+            initStringInfo(agg_key);
+            appendStringInfoString(agg_key, strVal(function));
+            appendStringInfoString(agg_key, ".");
```

Comment:
Just to check my understanding, does the Python FDW get a dict of `{"functionname.colname": {"function": "functionname", "column": "colname"}}` and is then expected to return a surrogate `functionname.colname` column in its response? e.g. https://github.com/splitgraph/postgres-elasticsearch-fdw/pull/1/files#diff-45ed0634a3ed30705f0b30dce58a096decc81bdf04af2df3906bc56d692c3de4R88-R92

Reply:
Yes, [that](https://github.com/splitgraph/postgres-elasticsearch-fdw/pull/1/files#diff-1bb6d76ca1dfdb92f1d20baf10faf031249c442578824c914dd5c0566399ae37R322-R334) is correct.